packages/@sanity/form-builder/src/sanity/legacyPartImplementations/__tests__/form-builder.test.tsx

```diff
@@ -0,0 +1,376 @@
+        {name: 'hidden', type: 'string', hidden: callbackFn},
+
+        /* String fields with readOnly */
+        {
+          name: 'isReadOnly',
+          type: 'string',
+          readOnly: true,
+        },
+        {
+          name: 'isNotReadOnly',
+          type: 'string',
+          readOnly: false,
+        },
+        {
+          name: 'isReadOnlyCallback',
+          type: 'string',
+          readOnly: readOnlyCallbackFn(true),
+        },
+        {
+          name: 'isNotReadOnlyCallback',
+          type: 'string',
+          readOnly: readOnlyCallbackFn(false),
+        },
+        {
+          name: 'isReadOnlyWithDocument',
+          type: 'string',
+          readOnly: readOnlyCallbackFnDocumentTrue,
+        },
+        {
+          name: 'isNotReadOnlyWithDocument',
+          type: 'string',
+          readOnly: readOnlyCallbackFnDocumentFalse,
+        },
+
+        /* String fields with hidden */
+        {
+          name: 'isHidden',
+          type: 'string',
+          hidden: true,
+        },
+        {
+          name: 'isNotHidden',
+          type: 'string',
+          hidden: false,
+        },
+        {
+          name: 'isHiddenCallback',
+          type: 'string',
+          hidden: hiddenCallbackFn(true),
+        },
+        {
+          name: 'isNotHiddenCallback',
+          type: 'string',
+          hidden: hiddenCallbackFn(false),
+        },
+        {
+          name: 'isHiddenWithDocument',
+          type: 'string',
+          hidden: hiddenCallbackFnDocumentTrue,
+        },
+        {
+          name: 'isNotHiddenWithDocument',
+          type: 'string',
+          hidden: hiddenCallbackFnDocumentFalse,
+        },
+        /* Boolean field with readOnly */
+        {
+          name: 'readOnlyBoolean',
+          type: 'boolean',
+          readOnly: true,
+        },
+        /* Image and file fields with readOnly */
+        {
+          name: 'readOnlyImage',
+          type: 'image',
+          readOnly: readOnlyCallbackFn(true),
+        },
+        {
+          name: 'readOnlyFile',
+          type: 'file',
+          readOnly: readOnlyCallbackFn(true),
+        },
+      ],
+    },
+  ],
+})
+
+const dummyDocument = {
+  _createdAt: '2021-11-04T15:41:48Z',
+  _id: 'drafts.10053a07-8647-4ebd-9d1d-33a512d30d3a',
+  _rev: '5hb8s6-k75-ip4-4bq-5ztbf3fbx',
+  _type: 'conditionalFieldsTest',
+  _updatedAt: '2021-11-05T12:34:29Z',
+  fieldWithObjectType: {
+    field1: 'dqwdw',
+    field2: 'vdsvdsvdsnkjnjkk',
+    roleConditionField: 'dvsvd',
+  },
+  isPublished: true,
+  readOnlyIfTitleIsReadOnly: 'dweefbfdssnflewnfklfwnlekfss',
+  readOnlyTestTitle: 'read only',
+  hiddenTestTitle: 'hide me',
+}
+
+const DEFAULT_PROPS = {
+  onFocus: jest.fn(),
+  onBlur: jest.fn(),
+  onChange: jest.fn(),
+  markers: [],
+  level: 0,
+  presence: [],
+}
+
+const documentType = schema.get('book')
+
+const FormBuilderTester = React.forwardRef(function FormBuilderTester() {
```

Comment:
we have created some helper components that you can use instead of having to write all of this in your test, you can look at the tests in our branch to see how we used it! 👍

Reply:
aha never mind, this is old, looks like your branch isn't using the latest from our branch...

Reply:
I actually think this is another of your files that I'm not sure why it's here 😅 
It was created by you, unless it was done in pair programming with @hermanwikner ?